php/libraries/SinglePointLogin.class.inc

```diff
@@ -305,60 +305,70 @@ class SinglePointLogin
                   GROUP BY UserID";
         $row   = $DB->pselectRow($query, array('username' => $username));
 
-        if (($row['User_count'] ?? 0) == 1) {
-            // validate passsword
-            $oldhash = $row['Password_hash'];
-            if (password_verify($password, $oldhash)) {
-                if (password_needs_rehash($oldhash, PASSWORD_DEFAULT)) {
-                    $user =& User::factory($username);
-                    $user->updatePassword($password);
-                }
-
-                if ($row['Active'] == 'N'
-                    || $this->disabledDueToInactivity($username, 365)
-                ) {
-                    $this->_lastError = "Your account has been deactivated."
-                        . " Please contact your project administrator to"
-                        . " reactivate this account.";
-                    $this->insertFailedDetail(
-                        "user account not active",
-                        $setArray
-                    );
+        if (intval($row['User_count'] ?? 0) !== 1) {
+            /* Display generic failure message to front-end so that an
+             * attacker will have a harder time knowing the difference
+             * between a bad username and a bad password.
+             */
+            $this->_lastError = "Incorrect username or password.";
+            return false;
+        }
+        // validate passsword
+        $oldhash = $row['Password_hash'];
+        if (!password_verify($password, $oldhash)) {
+            // Display generic failure message, as above.
+            $this->_lastError = "Incorrect username or password.";
+            $this->insertFailedDetail(
+                "Bad password",
+                $setArray
+            );
+            return false;
+        }
 
-                    return false;
-                }
+        if (password_needs_rehash($oldhash, PASSWORD_DEFAULT)) {
+            $user =& User::factory($username);
```

Comment:
Should it be `\User::` or `User::`?

Reply:
It should probably be `\User::` but I just moved that line from elsewhere in the old code.

Reply:
If the file's not in a namespace it doesn't really make a difference (but might be a good idea to include the `\` for consistency across files..)